### PR TITLE
Switch playground to engine patch files

### PR DESCRIPTION
**Pull request by loic-sharma**
Updates the playground to use patch files instead of an engine fork.

The `patches/001-Add-multi-view-Flutter-Windows-C++-APIs.patch` patch file updates the Flutter Windows `FlutterViewController` and allows creating multiple windows on Windows. See: https://github.com/flutter/flutter/issues/143767#issuecomment-2050741185

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -19,7 +19,7 @@ The demos in this repository run on the following platforms:
 | Platform | Support                                               |
 |----------|-------------------------------------------------------|
 | macOS    | ⚠️ with custom engine branch (see instructions below) |
-| Windows  | ⚠️ with custom engine branch (see instructions below) |
+| Windows  | ⚠️ with engine patch (see instructions below)         |
```
Comment (loic-sharma) [inline README.md:22]: ⚠️ macOS multi-view rendering does not work on latest engine main yet! This won't work until @dkwingsmt's latest work lands.

File: README.md [hunk 2]
```diff
@@ -36,11 +36,9 @@ git checkout master
 
 2. Set up the Engine development environment: See [the wiki page](https://github.com/flutter/flutter/wiki/Setting-up-the-Engine-development-environment).
 
-3. In **the engine repo**, add the prototype remote and switch to the prototype branch:
+3. In **the engine repo**, apply this playground's patches:
 ```
-git remote add dkwingsmt https://github.com/dkwingsmt/engine/
-git fetch dkwingsmt fffa6e19f5b71bedd55ac3a7fa6acfd85eda5983
-git checkout fffa6e19f5b71bedd55ac3a7fa6acfd85eda5983
+git apply /path/to/playground/patches/001-Add-multi-view-Flutter-Windows-C++-APIs.patch
```
Comment (goderbauer) [inline README.md:41]: Should the instructions here say something like: "For Windows use this patch:" - since macOS will likely be a different patch file?

File: README.md [hunk 3]
```diff
@@ -36,11 +36,9 @@ git checkout master
...
 ```
 
 4. In **this repo**, edit `pubspec.yaml`, and add the following dependency override to use the custom `dart:ui` library:
```
Comment (goderbauer) [inline README.md]: Also, while you're in here: Could you delete point 4? This is no longer necessary since the patch doesn't change any `dart:ui` API.

**Review by goderbauer — APPROVED** [review]
LGTM

File: README.md (hunk 3 above)
Comment (loic-sharma) [reply README.md]: Ah yup, good point!

Comment (loic-sharma) [reply README.md]: Removed

File: README.md (hunk 2 above)
Comment (loic-sharma) [reply README.md:41]: Updated! We can reconsider this if/when another platform also needs patch files.

Comment (dkwingsmt) [reply README.md:41]: This feels like you have to clone the engine repo to make the app work. In fact, to make the app download the files from the custom engine, I suppose the engine has to be fully compiled, which is not very available to many developers.

On the other hand, the developer can also simply override the files in the `ephemeral` directory, right? Although this means that they'll have to repatch it every time the engine is updated. Should we mention this too? Or is there a better way?

Comment (loic-sharma) [reply README.md:41]: Are you asking because you'd like Flutter macOS to use the "vanilla" engine without needing to a local compilation? Perhaps it would be better to have separate instructions for platforms that can use the "vanilla" engine and platforms that need a "custom" engine with patches.

To answer your question directly, yup the `ephemeral` directory approach could work as well. However, the instructions on how to maintain that are a little grosser. Also, the ergonomics there are a bit tricky - you need to build once using the `flutter` tool and then drop down to native tooling for subsequent builds.

Personally, I think the patch file approach is the best for Windows. It is a workflow that a Flutter engine developer is familiar with. Flutter Windows multi-view isn't productionized yet, you currently still need Flutter engine expertise to use it.

**Comment by wanjm** [comment]
@loic-sharma  @dkwingsmt  could you please list the steps of "ephemeral directory approach",thanks

**Comment by loic-sharma** [comment]
@wanjm Some background: Flutter Windows exposes a C API. The C++ API that Flutter Windows apps use is C++ code that's injected into the app's "ephemeral directory" by the Flutter tool each time you `flutter run` or `flutter build windows`. This C++ API just "wraps" the Flutter Windows C API with higher-level C++ abstractions.

Flutter Window's multi-view support is still experimental, so the C API that supports creating multiple views isn't in Flutter Windows's C header file. If you're willing to be hacky, you can edit Flutter Windows's C header file and C++ wrapper to use this experimental API. Note that this API isn't meant to be used yet: we may make breaking changes to it at any time. I would only use this to work on Flutter Windows's multi-window support.

I haven't tested this, but these steps should work:

1. Run `flutter build windows` in your app to add Flutter Windows's client wrapper code to your app.
2. After this step **do not** use `flutter build windows` or `flutter run`. This will reset your ephemeral directory and undo any changes you have done.
3. Update the ephemeral directory to use the experimental multi-view APIs:

   ```
      diff --git a/windows/flutter/ephemeral/cpp_client_wrapper/flutter_engine.cc b/windows/flutter/ephemeral/cpp_client_wrapper/flutter_engine.cc
      index 7860947..3cbe596 100644
      --- a/windows/flutter/ephemeral/cpp_client_wrapper/flutter_engine.cc
      +++ b/windows/flutter/ephemeral/cpp_client_wrapper/flutter_engine.cc
      @@ -63,7 +63,7 @@ bool FlutterEngine::Run(const char* entry_point) {
       }
       
       void FlutterEngine::ShutDown() {
      -  if (engine_ && owns_engine_) {
      +  if (engine_) {
           FlutterDesktopEngineDestroy(engine_);
         }
         engine_ = nullptr;
      @@ -113,8 +113,7 @@ std::optional<LRESULT> FlutterEngine::ProcessExternalWindowMessage(
         return std::nullopt;
       }
       
      -FlutterDesktopEngineRef FlutterEngine::RelinquishEngine() {
      -  owns_engine_ = false;
      +FlutterDesktopEngineRef FlutterEngine::engine() const {
         return engine_;
       }
       
      diff --git a/windows/flutter/ephemeral/cpp_client_wrapper/flutter_view_controller.cc b/windows/flutter/ephemeral/cpp_client_wrapper/flutter_view_controller.cc
      index 98c65e1..1c1b51e 100644
      --- a/windows/flutter/ephemeral/cpp_client_wrapper/flutter_view_controller.cc
      +++ b/windows/flutter/ephemeral/cpp_client_wrapper/flutter_view_controller.cc
      @@ -12,9 +12,32 @@ namespace flutter {
       FlutterViewController::FlutterViewController(int width,
                                                    int height,
                                                    const DartProject& project) {
      +  FlutterDesktopViewControllerProperties properties = {};
      +  properties.width = width;
      +  properties.height = height;
      +
         engine_ = std::make_shared<FlutterEngine>(project);
      -  controller_ = FlutterDesktopViewControllerCreate(width, height,
      -                                                   engine_->RelinquishEngine());
      +  controller_ =
      +      FlutterDesktopEngineCreateViewController(engine_->engine(), &properties);
      +  if (!controller_) {
      +    std::cerr << "Failed to create view controller." << std::endl;
      +    return;
      +  }
      +  view_ = std::make_unique<FlutterView>(
      +      FlutterDesktopViewControllerGetView(controller_));
      +}
      +
      +FlutterViewController::FlutterViewController(
      +    int width,
      +    int height,
      +    std::shared_ptr<FlutterEngine> engine) {
      +  FlutterDesktopViewControllerProperties properties = {};
      +  properties.width = width;
      +  properties.height = height;
      +
      +  engine_ = std::move(engine);
      +  controller_ =
      +      FlutterDesktopEngineCreateViewController(engine_->engine(), &properties);
         if (!controller_) {
           std::cerr << "Failed to create view controller." << std::endl;
           return;
      diff --git a/windows/flutter/ephemeral/cpp_client_wrapper/include/flutter/flutter_engine.h b/windows/flutter/ephemeral/cpp_client_wrapper/include/flutter/flutter_engine.h
      index 0369db3..89ca2d1 100644
      --- a/windows/flutter/ephemeral/cpp_client_wrapper/include/flutter/flutter_engine.h
      +++ b/windows/flutter/ephemeral/cpp_client_wrapper/include/flutter/flutter_engine.h
      @@ -98,11 +98,8 @@ class FlutterEngine : public PluginRegistry {
         // For access to the engine handle.
         friend class FlutterViewController;
       
      -  // Gives up ownership of |engine_|, but keeps a weak reference to it.
      -  //
      -  // This is intended to be used by FlutterViewController, since the underlying
      -  // C API for view controllers takes over engine ownership.
      -  FlutterDesktopEngineRef RelinquishEngine();
      +  // Get the handle for interacting with the C API's engine reference.
      +  FlutterDesktopEngineRef engine() const;
       
         // Handle for interacting with the C API's engine reference.
         FlutterDesktopEngineRef engine_ = nullptr;
      @@ -110,9 +107,6 @@ class FlutterEngine : public PluginRegistry {
         // Messenger for communicating with the engine.
         std::unique_ptr<BinaryMessenger> messenger_;
       
      -  // Whether or not this wrapper owns |engine_|.
      -  bool owns_engine_ = true;
      -
         // Whether |Run| has been called successfully.
         //
         // This is used to improve error messages. This can be false while the engine
      diff --git a/windows/flutter/ephemeral/cpp_client_wrapper/include/flutter/flutter_view_controller.h b/windows/flutter/ephemeral/cpp_client_wrapper/include/flutter/flutter_view_controller.h
      index 4007534..b26e017 100644
      --- a/windows/flutter/ephemeral/cpp_client_wrapper/include/flutter/flutter_view_controller.h
      +++ b/windows/flutter/ephemeral/cpp_client_wrapper/include/flutter/flutter_view_controller.h
      @@ -32,6 +32,16 @@ class FlutterViewController {
         // |dart_project| will be used to configure the engine backing this view.
         FlutterViewController(int width, int height, const DartProject& project);
       
      +  // Creates a FlutterView that can be parented into a Windows View hierarchy
      +  // either using HWNDs.
      +  //
      +  // This creates the view on an existing FlutterEngine.
      +  //
      +  // |dart_project| will be used to configure the engine backing this view.
      +  FlutterViewController(int width,
      +                        int height,
      +                        std::shared_ptr<FlutterEngine> engine);
      +
         virtual ~FlutterViewController();
       
         // Prevent copying.
      @@ -44,6 +54,9 @@ class FlutterViewController {
         // Returns the engine running Flutter content in this view.
         FlutterEngine* engine() const { return engine_.get(); }
       
      +  // Returns the engine running Flutter content in this view.
      +  std::shared_ptr<FlutterEngine> shared_engine() const { return engine_; }
      +
         // Returns the view managed by this controller.
         FlutterView* view() const { return view_.get(); }
       
      diff --git a/windows/flutter/ephemeral/flutter_windows.dll.exp b/windows/flutter/ephemeral/flutter_windows.dll.exp
      index 9bd1773..5b81194 100644
      Binary files a/windows/flutter/ephemeral/flutter_windows.dll.exp and b/windows/flutter/ephemeral/flutter_windows.dll.exp differ
      diff --git a/windows/flutter/ephemeral/flutter_windows.h b/windows/flutter/ephemeral/flutter_windows.h
      index 80d7876..d7b2a30 100644
      --- a/windows/flutter/ephemeral/flutter_windows.h
      +++ b/windows/flutter/ephemeral/flutter_windows.h
      @@ -70,6 +70,15 @@ typedef struct {
       
       } FlutterDesktopEngineProperties;
       
      +// Properties for configuring a Flutter view controller.
      +typedef struct {
      +  // The view's initial width.
      +  int width;
      +
      +  // The view's initial height.
      +  int height;
      +} FlutterDesktopViewControllerProperties;
      +
       // ========== View Controller ==========
       
       // Creates a view that hosts and displays the given engine instance.
      @@ -165,6 +174,22 @@ FLUTTER_EXPORT bool FlutterDesktopEngineDestroy(FlutterDesktopEngineRef engine);
       FLUTTER_EXPORT bool FlutterDesktopEngineRun(FlutterDesktopEngineRef engine,
                                                   const char* entry_point);
       
      +// Creates a view for the given engine.
      +//
      +// The |engine| will be started if it is not already running.
      +//
      +// The caller owns the returned reference, and is responsible for calling
      +// |FlutterDesktopViewControllerDestroy|. Returns a null pointer in the event of
      +// an error.
      +//
      +// Unlike |FlutterDesktopViewControllerCreate|, this does *not* take ownership
      +// of |engine| and |FlutterDesktopEngineDestroy| must be called to destroy
      +// the engine.
      +FLUTTER_EXPORT FlutterDesktopViewControllerRef
      +FlutterDesktopEngineCreateViewController(
      +    FlutterDesktopEngineRef engine,
      +    const FlutterDesktopViewControllerProperties* properties);
      +
       // DEPRECATED: This is no longer necessary to call, Flutter will take care of
       // processing engine messages transparently through DispatchMessage.
       //
      
       ```

5. Open `./build/windows/<architecture>/<app name>.sln` using Visual Studio
6. Use Visual Studio to launch your app (don't use `flutter run` or `flutter build windows` as that'll reset your ephemeral directory!)

**Comment by loic-sharma** [comment]
I created a sample repo that launches multiple windows on Windows: https://github.com/loic-sharma/windows_multi_window

This requires Flutter's `master` channel but doesn't require a custom engine build or patching your ephemeral directory.

**Comment by wanjm** [comment]
@loic-sharma  thank you so much!
